libcxx/test/support/exception_test_helpers.h

```diff
@@ -0,0 +1,87 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef EXCEPTION_TEST_HELPERS_H
+#define EXCEPTION_TEST_HELPERS_H
+
+#include "count_new.h"
+
+template <class T>
+struct throwing_allocator {
+  using value_type      = T;
+  using is_always_equal = std::false_type;
```

Comment:
This isn't telling the truth -- why? This would need a comment.

Reply:
This class was named `Allocator<T>` which sounds a general allocator class with no indication of whether it may throw exception. Renaming it from `Allocator` to `throwing_allocator` makes its purpose to throw exceptions more explicit.

Reply:
Thank you for this comment, which motivated me to think more deeply about the issue. I just realized that I must have misunderstood the question initially. Now I believe the question pertains to the line `using is_always_equal = std::false_type;`, which appears counter-intuitive for a stateless allocator class. 

This line was in the original class `Allocator` before my refactor, my best guess is that the original author intended to use this class to test the exception safety for the constructor `vector(vector&&, const allocator_type&)` in `vector/vector.cons/exceptions.pass.cpp`: 

```cpp
try { // Throw in vector(vector&&, const allocator_type&) from type
    std::vector<ThrowingT, Allocator<ThrowingT> > vec(Allocator<ThrowingT>(false));
    int throw_after = 1;
    vec.emplace_back(throw_after);
    std::vector<ThrowingT, Allocator<ThrowingT> > vec2(std::move(vec), Allocator<ThrowingT>(false));
  } catch (int) {
  }
  check_new_delete_called();
```

This test aims to verify the exception safety for the constructor while performing an O(n) element-wise move (rather than the O(1) resource stealing), which requires the instances of the Allocator class to compare unequal. However, merely  setting `Allocator::is_always_equal::value` to false does not ensure that the instances of this class are unequal, which renders the above test actually testing nothing, as you also confirmed [here](https://github.com/llvm/llvm-project/pull/116449#pullrequestreview-2462660650). 
 

Beyond this scenario, I couldn't find any reason why we might need this member type. I believe we can safely remove it. Would you suggest I remove it in this PR or in my next one while rebasing?

Reply:
The above test has been replaced by the following test:

```cpp
try { // Throw in vector(vector&&, const allocator_type&) from type during element-wise move
  std::vector<throwing_t, test_allocator<throwing_t> > vec(test_allocator<throwing_t>(1));
  int throw_after = 10;
  throwing_t v(throw_after);
  vec.insert(vec.end(), 6, v);
  std::vector<throwing_t, test_allocator<throwing_t> > vec2(std::move(vec), test_allocator<throwing_t>(2));
} catch (int) {
}
check_new_delete_called();
```

This test now uses  `test_allocator<T>`, which checks equality between the allocator instances  based on the `data_` field. In the above test, the data fields 1 and 2 make the two allocators unequal, thus ensuring element-wise move. 

The original test use a stateless allocator type `Allocator` (now renamed to `throwing_allocator`), which is not suitable for the above purpose. Also, the fact that a stateless allocator type has a false `is_always_equal` value sounds a bit counter-intuitive to me. That's why I removed the member type in my later PR. Now `locator_traits<throwing_allocator>::is_always_equal` becomes true and this is consistence with its `operator==` which always returns true.